Mage.Sets/src/mage/cards/f/FirestormPhoenix.java

```diff
@@ -0,0 +1,203 @@
+	        	controller.moveCards(permanent, Zone.HAND, source, game);
+	        
+	            return true;
+	        }
+	    }
+	    return false;
+	}
+	
+    @Override
+    public boolean checksEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.ZONE_CHANGE;
+    }
+	
+    @Override
+    public boolean applies(GameEvent event, Ability source, Game game) {
+    	return source.getSourceId().equals(event.getTargetId()) 
+        && ((ZoneChangeEvent) event).isDiesEvent();
+    }
+}
+
+class FirestormPhoenixRevealEffect extends ContinuousEffectImpl {
+	
+	private Player controller;
+	private Cards revealedCards;
+	private String cardName;
+
+	FirestormPhoenixRevealEffect(Player tmp_controller, Cards tmp_revealedCards, String tmp_cardName) {
+        super(Duration.UntilYourNextTurn, Outcome.Detriment);
+        controller = tmp_controller;
+        revealedCards = tmp_revealedCards;
+        cardName = tmp_cardName;
+        staticText = "Until that player’s next turn, that player plays with that card revealed in their hand and can’t play it";
+    }
+
+    private FirestormPhoenixRevealEffect(final FirestormPhoenixRevealEffect effect) {
+        super(effect);
+        this.controller = effect.controller;
+        this.revealedCards = effect.revealedCards;
+        this.cardName = effect.cardName;
+    }
+
+    @Override
+    public FirestormPhoenixRevealEffect copy() {
+        return new FirestormPhoenixRevealEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+    	return false;
+    }
+
+    @Override
+    public boolean apply(Layer layer, SubLayer sublayer, Ability source, Game game) {
+    	for(UUID permID : revealedCards) {
+    		if(game.getState().getZone(permID) == Zone.HAND) {
+        		controller.revealCards(cardName, revealedCards, game, false);
+        	} else {
+        		discard();
+        	}
+    	}
+		return true;
+    }
+
+    @Override
+    public boolean hasLayer(Layer layer) {
+        return true;
+    }
+}
+
+class FirestormPhoenixRestrictEffect extends ContinuousRuleModifyingEffectImpl {
+	
+	String cardName;
+	private MageObjectReference mor;
+	
+	public FirestormPhoenixRestrictEffect(Permanent permanent, Game game) {
+        super(Duration.UntilYourNextTurn, Outcome.Detriment);
+        mor = new MageObjectReference(permanent, game);
+        staticText = "Until that player’s next turn, that player plays with that card revealed in their hand and can’t play it";
+    }
+
+    private FirestormPhoenixRestrictEffect(final FirestormPhoenixRestrictEffect effect) {
+        super(effect);
+        this.mor = effect.mor;
+    }
+
+    @Override
+    public FirestormPhoenixRestrictEffect copy() {
+        return new FirestormPhoenixRestrictEffect(this);
+    }
+    
+    @Override
+    public boolean checksEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.CAST_SPELL_LATE;
```

Comment:
I'm not sure it's really needed to check on CAST_SPELL_LATE here. UX is better checking CAST_SPELL. It's possible there's some contrived interaction that lets you cast it with different characteristics such that it wouldn't be affected by this effect, but I can't think of what would do it - the restriction is pretty fundamentally "can't play that card"

Reply:
Yeah, I couldn't think of an interaction where this would really matter either, but was trying to cover all my bases. I'll switch this to CAST_SPELL